### DISCUSSION TITLE
PhysX Linker Error

### DISCUSSION THREAD

**Discussion by ZaffyPuck** [NVIDIA-Omniverse/PhysX#106] [Q&A] [answered]
Hello!

I am incorporating PhysX libraries into my own graphics engine using DX11. I unfortunately seem to be running into some linking errors and can't seem to get it working. With lack of similar issues and solutions available, I figured I could ask here for help. The errors seem to be signifying that there is no definition of the PxDefaultErrorCallback. I have attached a picture for reference. I am trying to initialize the physics and have been following some of the "hello world" snippets available. I hope someone can help, let me know if any more information is necessary.

![Screenshot 2023-03-01 221023](https://user-images.githubusercontent.com/56975883/222321279-e3e3bec6-1f5b-4b32-bf8b-744cb632e606.png)

**Comment by ayoub-belarbi** [comment, marked as answer]
Hello @ZaffyPuck,

Did you link against PhysXExtensions? If yes, did you include PxConfig.h header that will contain PhysX configuration defines like PX_PHYSX_STATIC_LIB?
Alternatively, You can also directly add #define PX_PHYSX_STATIC_LIB as a first thing or add PX_PHYSX_STATIC_LIB to your preprocessor defines in any other way. 

Cheers,
Ayoub

**Reply by ZaffyPuck** [reply]
I attached a picture of my lib imports as well as the includes I am currently using. I am not currently including the PxConfig.h file, but it didn't seem to make a difference when I did so, nor does defining PX_PHYSX_STATIC_LIB. Perhaps I am not linking correctly? Though I don't have any lib errors.

![Screenshot 2023-03-02 110153](https://user-images.githubusercontent.com/56975883/222483792-c5404b0b-41dc-4305-a0bc-caf2a8d8013e.png)

**Reply by ZaffyPuck** [reply]
I probably should have specified, I am trying to link the PhysX library through .dlls rather than the .libs. It seems that there is no PhysXExtensions .dll, which is why it is throwing the errors in the initial post. I am wondering if the library is even capable of being imported in such a way, because they way I have DX11 linked requires the Runtime Library configuration to be Multi-threaded Debug DLL (/MDd). Let me know if you have any ideas.

**Reply by ayoub-belarbi** [reply]
Ah ok now it's clear.
You can switch to dynamic runtime usage by toggling the switch: NV_USE_STATIC_WINCRT.
For example, if you are using vc16, switch this to False: https://github.com/NVIDIA-Omniverse/PhysX/blob/release/104.2/physx/buildtools/presets/public/vc16win64.xml#L8

same applies if you use vc17, you need to change it in vc17win64
Please have a look at: https://nvidia-omniverse.github.io/PhysX/physx/5.1.3/docs/BuildingWithPhysX.html

thanks

**Reply by ZaffyPuck** [reply]
Hey Ayoub,

I was able to get PhysX compiled with your advice! So big thanks for that. Would you happen to know if there is anything more that needs to be done to link source files to the project as well? The header files cant seem to find the correct function definitions. For example, "PxCreateFoundation" can't find where its definition lies, which is in the "FdFoundation.cpp" file. I have all the PhysX source files linked in project properties>>VC++ Directories>>Source Directories, but this doesn't seem to be working. Just curious if you have any advice.

**Comment by ayoub-belarbi** [comment]
Hello again,
Glad to hear that it worked out :)
I really can't help you with your last question since it depends on how you are setting up the project, your dev environemnt etc and it's a more general C++/Visual Studio question rather than PhysXSDK.

**Reply by ZaffyPuck** [reply]
I got it working, thanks!
